File: 2025-05-07/dz1.py
#!/usr/bin/python3

import random

# Функция ввода числа с проверкой
def Num(txt, n1, n2):
    if n1 > n2: n1, n2 = n2, n1
    _ = input(f'{txt}: ')
    if _.isnumeric() and n1 <= int(_) <=n2: return int(_)
    else: print(f'Ошибка ввода! Необходимо ввести число от {n1} до {n2}!\n'); return Num(txt, n1, n2)

def Sort(mass, x):
    for i in range(x):
       for j in range(i, x):
          if mass[i] > mass[j]: mass[i], mass[j] = mass[j], mass[i]
    return mass

rnd = [random.randint(-100, 100) for i in range(30)]

# Задание 1
print(f'Первоначальный список: {rnd}')
if sum(rnd) / len(rnd) > 0 : print('Больше нуля!'); Sort(rnd, len(rnd) // 3 * 2)
else: print('Меньше нуля!'); Sort(rnd, len(rnd) // 3)
print(f'Конечный список: {rnd}')

# Задание 2
mass = []
while len(mass) < 10: mass.append(Num(f'Введите оценку {len(mass) + 1}', 1, 12))
print(f'\nОценки: {mass}')
mass[Num('Какую оценку заменить (1 - 10)', 1, 10) -1] = Num('\nВведите новую оценку', 1, 12)
print(f'Новые Оценки: {mass}')
if sum(mass) > 70: print('Стипендия будет!')
else: print('Стипендии не будет :(')
mass.sort()
print(f'Отсортированные оценки {mass}')

# Задание 3
rnd = [random.randint(-100, 100) for i in range(30)]
print(f'Первоначальный список: {rnd}')
i = 1
while i != 0:
    i = 0
    for j in range(len(rnd) - 1):
      if rnd[j] > rnd[j+1]: rnd[j], rnd[j+1] = rnd[j+1], rnd[j]; i += 1
print(f'Отсортированный список: {rnd}')

File: 2025-05-07/dz1.go
package main

import (
    "fmt"
    "math/rand"
    "strconv"
)

// Функция ввоад числа с проверкой
func Num(txt string, n1, n2 int) int {
    var p string
    fmt.Printf("%s: ", txt)
    fmt.Scanln(&p)
    y, err := strconv.Atoi(p)
    if err == nil && y >= n1 && y <= n2 { return y } else { fmt.Printf("Ошибка ввода! Необходимо ввести число от %d до %d!\n", n1, n2); return Num(txt, n1, n2)}
}

func Summ(a []int) int {
    i :=0
    for _, j := range a  { i += j }
    return i
}

func Sort(a []int, x int) []int {
    for i := 0; i <  x; i++ {
       for j := i; j < x; j ++ { if a[i] > a[j] { a[i], a[j] = a[j], a[i] } }
    }
    return a
}

func main () {
    mass := []int {}
    // Задание 1
    for len(mass) < 30 { mass = append(mass, rand.Intn(200) - 100) }
    fmt.Println("ЗАДАНИЕ 1: Необходимо отсортировать первые две трети списка в порядке возрастания, если среднее арифметическое всех элементов больше нуля; иначе — лишь первую треть. Остальную часть списка не сортировать, а расположить в обратном порядке.\n\nРЕШЕНИЕ:")
    fmt.Printf("Первоначальный список (%d, элементов): %d\n", len(mass), mass)
    if Summ(mass) / len(mass) > 0 { fmt.Printf("Средняя арифметическое: %d\nСписок с отсортированными элементами от 0 до %d: %d\n", Summ(mass)/len(mass), len(mass) / 3 * 2 - 1, Sort(mass, len(mass) / 3 * 2))
    } else { fmt.Printf("Средняя арифметическое: %d\nСписок с отсортированными элементами от 0 до %d: %d\n", Summ(mass)/len(mass), len(mass) / 3 - 1, Sort(mass, len(mass) / 3))}

    // Задание 2
    mass = nil
    fmt.Println("\n\nЗАДАНИЕ 2: Написать программу «успеваемость». Пользователь вводит 10 оценок студента. Оценки от 1 до 12. Реализовать меню для пользователя:\n - Вывод оценок (вывод содержимого списка);\n - Пересдача экзамена (пользователь вводит номер элемента списка и новую оценку);\n - Выходит ли стипендия (стипендия выходит, если средний бал не ниже 10.7);\n - Вывод отсортированного списка оценок: по возрастанию или убыванию.\n\nРЕШЕНИЕ:")
    for len(mass) < 10 { mass = append(mass, Num(fmt.Sprintf("Введите оценку %d", len(mass) + 1), 1, 12)) }
    fmt.Printf("Оценки: %d\n", mass)
    mass[Num("Какую оценку заменить (1  - 10)", 1, 10) -1] = Num("Введите новую оценку", 1, 12)
    fmt.Printf("Новые Оценки: %d\nСредний бал: %.2f. ", mass, float32(Summ(mass)) / float32(len(mass)))
    if float32(Summ(mass)) / float32(len(mass)) > 10.7 { fmt.Println("Стипендия будет!") } else { fmt.Println("Стипендии не будет :(") }
    fmt.Printf("Отсортированные оценки: %d\n", Sort(mass, len(mass)))

    // Задание 3
    mass = nil
    for len(mass) < 30 { mass = append(mass, rand.Intn(200) - 100) }
    fmt.Println("\n\nЗАДАНИЕ 3: Написать программу, реализующую сортировку списка методом усовершенствованной сортировки пузырьковым методом. Усовершенствование состоит в том, чтобы анализировать количество перестановок на каждом шагу, если это количество равно нулю, то продолжать сортировку нет смысла — список отсортирован.\n\nРЕШЕНИЕ:")
    fmt.Printf("Первоначальный список: %d\n", mass)
    i := 1
    for  i != 0 {
       i = 0
      for j :=0; j < len(mass) - 1; j++ { if mass[j] > mass[j+1] { mass[j], mass[j + 1] = mass[j + 1], mass[j]; i += 1 } }
    }
    fmt.Printf("Отсортированный список: %d\n", mass)
}